contracts/contracts/MessageProcessor.sol

```diff
@@ -0,0 +1,250 @@
+    uint256 public currentMessageBatchIndex;
+   // The commitment to the state and ballot roots
+    uint256 public sbCommitment;
+
+    Verifier public verifier;
+
+    constructor(Verifier _verifier) {
+        verifier = _verifier;
+    }
+
+    modifier votingPeriodOver(Poll _poll) {
+        (uint256 deployTime, uint256 duration) = _poll
+            .getDeployTimeAndDuration();
+        // Require that the voting period is over
+        uint256 secondsPassed = block.timestamp - deployTime;
+        require(secondsPassed > duration, ERROR_VOTING_PERIOD_NOT_PASSED);
+        _;
+    }
+
+    /*
+     * Update the Poll's currentSbCommitment if the proof is valid.
+     * @param _poll The poll to update
+     * @param _newSbCommitment The new state root and ballot root commitment
+     *                         after all messages are processed
+     * @param _proof The zk-SNARK proof
+     */
+    function processMessages(
+        Poll _poll,
+        uint256 _newSbCommitment,
+        uint256[8] memory _proof
+    ) public onlyOwner votingPeriodOver(_poll) {
+        // There must be unprocessed messages
+        require(!processingComplete, ERROR_NO_MORE_MESSAGES);
+
+        // The state AccQueue must be merged
+        require(_poll.stateAqMerged(), ERROR_STATE_AQ_NOT_MERGED);
+
+        // Retrieve stored vals
+        (, , uint8 messageTreeDepth, ) = _poll.treeDepths();
+        (uint256 messageBatchSize, , ) = _poll.batchSizes();
+
+        AccQueue messageAq;
+        (, , messageAq, ) = _poll.extContracts();
+
+        // Require that the message queue has been merged
+        uint256 messageRoot = messageAq.getMainRoot(messageTreeDepth);
+        require(messageRoot != 0, ERROR_MESSAGE_AQ_NOT_MERGED);
+
+        // Copy the state and ballot commitment and set the batch index if this
+        // is the first batch to process
+        if (numBatchesProcessed == 0) {
+            uint256 currentSbCommitment = _poll.currentSbCommitment();
+            sbCommitment = currentSbCommitment;
+            (, uint256 numMessages) = _poll.numSignUpsAndMessages();
+            uint256 r = numMessages % messageBatchSize;
+
+            if (r == 0) {
+                currentMessageBatchIndex =
+                    (numMessages / messageBatchSize) *
+                    messageBatchSize;
+            } else {
+                currentMessageBatchIndex = numMessages;
+            }
+
+            if (currentMessageBatchIndex > 0) {
+                if (r == 0) {
+                    currentMessageBatchIndex -= messageBatchSize;
+                } else {
+                    currentMessageBatchIndex -= r;
+                }
+            }
+        }
+
+        bool isValid = verifyProcessProof(
+            _poll,
+            currentMessageBatchIndex,
+            messageRoot,
+            sbCommitment,
+            _newSbCommitment,
+            _proof
+        );
+        require(isValid, ERROR_INVALID_PROCESS_MESSAGE_PROOF);
+
+        {
```

Comment:
let's move state update to the top of the function. Will require caching the state variables as local variables so they can be used in the function body